modules/src/applications/transfer/context.rs

```diff
@@ -0,0 +1,348 @@
+use sha2::{Digest, Sha256};
+use subtle_encoding::hex;
+
+use super::error::Error as Ics20Error;
+use crate::applications::transfer::acknowledgement::Acknowledgement;
+use crate::applications::transfer::events::{AckEvent, AckStatusEvent, RecvEvent, TimeoutEvent};
+use crate::applications::transfer::packet::PacketData;
+use crate::applications::transfer::relay::on_ack_packet::process_ack_packet;
+use crate::applications::transfer::relay::on_recv_packet::process_recv_packet;
+use crate::applications::transfer::relay::on_timeout_packet::process_timeout_packet;
+use crate::applications::transfer::{DenomTrace, HashedDenom, IbcCoin, VERSION};
+use crate::core::ics04_channel::channel::{Counterparty, Order};
+use crate::core::ics04_channel::context::{ChannelKeeper, ChannelReader};
+use crate::core::ics04_channel::msgs::acknowledgement::Acknowledgement as GenericAcknowledgement;
+use crate::core::ics04_channel::packet::Packet;
+use crate::core::ics04_channel::Version;
+use crate::core::ics05_port::context::PortReader;
+use crate::core::ics24_host::identifier::{ChannelId, ConnectionId, PortId};
+use crate::core::ics26_routing::context::{ModuleOutputBuilder, OnRecvPacketAck};
+use crate::prelude::*;
+use crate::signer::Signer;
+
+pub trait Ics20Keeper:
+    ChannelKeeper + BankKeeper<AccountId = <Self as Ics20Keeper>::AccountId>
+{
+    type AccountId: ToString;
+
+    /// Sets a new {trace hash -> denom trace} pair to the store.
+    fn set_denom_trace(&mut self, denom_trace: &DenomTrace) -> Result<(), Ics20Error>;
+}
+
+pub trait Ics20Reader:
+    BankReader<AccountId = <Self as Ics20Reader>::AccountId>
+    + AccountReader<AccountId = <Self as Ics20Reader>::AccountId>
+    + ChannelReader
+    + PortReader
+{
+    type AccountId: ToString + TryFrom<Signer>;
+
+    /// get_port returns the portID for the transfer module.
+    fn get_port(&self) -> Result<PortId, Ics20Error>;
+
+    /// Returns the escrow account id for a port and channel combination
+    fn get_channel_escrow_address(
+        &self,
+        port_id: &PortId,
```

Comment:
ICS-20 indexes the escrow addresses only by channel id, whereas ibc-go uses channel+port ids. Since channel ids are already assumed to be unique, isn't the `port_id` redundant here? I think we should do as ICS-20 does and index only by channel id

Reply:
Seems like this is related to a vulnerability in the spec in the case of 'crossing-hellos' identified during an audit (see point 2 here https://github.com/cosmos/ibc-go/issues/65). There's an ibc-go PR for removing 'crossing-hellos' (https://github.com/cosmos/ibc-go/issues/1311), so it might be nice to raise this concern there and ask if there are any plans to backport this to ibc-go v2.

Reply:
seem the function `get_channel_escrow_address` for one chain have oneself implement.